mamba-org/mamba
# maint: Use Catch2 instead of doctest

**Why**: mostly because `doctest` doesn't seem to be well supported: https://github.com/doctest/doctest/issues/554
Last commit to `main` was **March 15, 2023**, to `dev` there were only 8 commits after that.
`Catch2` last commit was last week, and the project is incredibly popular.

Please, let me know what you think. If we decide to switch to Catch2, I will switch in all the repo.

My personal opinion: `Catch2` is easy to use, fast enough, well supported, and feels quite modern.

How to switch from `doctest` to `Catch2` (easy things):
- small changes to `cmake` for `find_package` and `target_link_libraries`
- `#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN` -> `#define CATCH_CONFIG_MAIN`
- `#include <doctest/doctest.h>` -> `#include <catch2/catch_all.hpp>`
- `SUBCASE` -> `SECTION`
- `CHECK` -> `REQUIRE`, `CHECK_FALSE` -> `REQUIRE_FALSE`
- `CHECK_EQ(a, b);` and so on -> `REQUIRE(a == b);` (can be almost done with regex replacement, a bit messy when there are commas in `a` or `b`)

A bit more complicated:
- `TEST_SUITE` - doesn't exist in Catch2, but we can use tags, a second arg to `TEST_CASE`. I also added brackets to tag, because this is how tags look like in Catch2.

Notes:
- `doctest is modeled after [Catch](https://github.com/catchorg/Catch2) and some parts of the code have been taken directly` (https://github.com/doctest/doctest), that's why the switch is relatively easy
- in places where `TEST_SUITE` was used, I added `namespace {}`, it allows to keep indentation the same so the diff is simple
- There is both `REQUIRE` and `CHECK` in Catch2, the difference is: The `REQUIRE` family of macros tests an expression and aborts the test case if it fails. The `CHECK` family are equivalent but execution continues in the same test case even if the assertion fails. I use `REQUIRE` to fail fast.
